test/persistence-hooks.suite.js

```diff
@@ -1828,6 +2068,442 @@ module.exports = function(dataSource, should) {
+
+        TestModel.replaceOrCreate(
+          { id: 'new-id', name: 'a name' },
+          function(err, instance) {
+            if (err) return done(err);
+
+            if (dataSource.connector.replaceOrCreate) {
+              // Atomic implementations of `replaceOrCreate`
+              observedContexts.should.eql(aTestModelCtx({
+                instance: {
+                  id: 'new-id',
+                  name: 'a name',
+                  extra: undefined
+                }
+              }));
+            } else {
+              // non-atomic implementation of `replaceOrCreate`
+              // will use `instance.save`
+              observedContexts.should.eql(aTestModelCtx({
+                instance: { id: 'new-id', name: 'a name', extra: undefined },
+                isNewInstance: true
+              }));
+            }
+            done();
+          });
+      });
+      
+      it('applies updates from `before save` hook on create', function(done) {
+        TestModel.observe('before save', function(ctx, next) {
+          ctx.instance.name = 'hooked';
+          next();
+        });
+
+        TestModel.replaceOrCreate(
+          { id: 'new-id', name: 'new name' },
+          function(err, instance) {
+            if (err) return done(err);
+            instance.name.should.equal('hooked');
+            done();
+          });
+      });
+      
+      it('validates model after `before save` hook on create', function(done) {
+        TestModel.observe('before save', invalidateTestModel());
+
+        TestModel.replaceOrCreate(
+          { id: 'new-id', name: 'new name' },
+          function(err, instance) {
+            (err || {}).should.be.instanceOf(ValidationError);
+            (err.details.codes || {}).should.eql({ name: ['presence'] });
+            done();
+          });
+      });
+      
+      it('triggers `persist` hook on create', function(done) {
+        TestModel.observe('persist', pushContextAndNext());
+
+        TestModel.replaceOrCreate(
+          { id: 'new-id', name: 'a name' },
+          function(err, instance) {
+            if (err) return done(err);
+
+            if (dataSource.connector.replaceOrCreate) {
+              observedContexts.should.eql(aTestModelCtx({
+                where: { id: 'new-id' },
+                data: { id: 'new-id', name: 'a name' },
+                currentInstance: {
+                  id: 'new-id',
+                  name: 'a name',
+                  extra: undefined
+                }
+              }));
+            } else {
+              observedContexts.should.eql(aTestModelCtx({
+                data: {
+                  id: 'new-id',
+                  name: 'a name'
+                },
+                isNewInstance: true,
+                currentInstance: {
+                  id: 'new-id',
+                  name: 'a name',
+                  extra: undefined
+                }
+              }));
```

Comment:
Oh I see, non-atomic implementation does not provide `ctx.where` because a new instance is being created, so there are not records to match `where` filter.

Can you please still apply the cleanup (add `expectedContext`) and add a comment explaining why `ctx.where` is not set in one of the branches/